substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageBFDNameProvider.java

```diff
@@ -375,22 +412,113 @@ public String bfdMangle(Member m) {
+         * result that it will demangle to Hello and can be used to replace later occurrences of the
+         * string Hello.
+         *
+         * When a sequence of simple names is mangled inside a namespace substitution bindings are
+         * recorded for each successive composite namespace prefix but not for the final symbol
+         * itself. For example, when method name Hello::main is mangled to N5Hello4main4 a single
+         * binding is recorded which demangles to Hello. If a class has been loaded by an
+         * application loader and, hence, has a name which includes a loader namespace prefix (to
+         * avoid the possibility of the same named class being loaded by two different loaders) then
+         * the method name, say AppCL504::Hello::main, would mangle to a namespace encoding with 3
+         * elements N8AppCL5045Hello4mainE which would introduce two bindings, the first demangling
+         * to AppCL504 and the second to AppCL504::Hello.
+         *
+         * A binding is also recorded whenever a pointer type is mangled. The bound symbol demangles
+         * to whatever the text decoded from the scope of the P followed by a '*' to translate it to
+         * a pointer type. For example, when the type in parameter list (Foo*) is encoded as P3Foo a
+         * binding is recorded for the pointer as well as for the type Foo. The first binding
+         * demangles to Foo. The second binding demangles to Foo*.
+         *
+         * n.b. repeated use of the pointer operator results in repeated bindings. So, if a
+         * parameter with type Foo** were to be encoded as PP3Foo three bindings would recorded
+         * which demangle to Foo, Foo* and Foo**. However, multiple indirections do not occur in
+         * Java signatures.
+         *
+         * n.b.b. repeated mentions of the same symbol would redundantly record a binding. For
+         * example if int Hello::compareTo(Hello*) were mangled to _ZN5Hello9compareToEJiP5Hello the
+         * resulting bindings would be S_ ==> Hello, S0_ ==> Hello::compareTo, S1_ ==> Hello and S2_
+         * ==> Hello* i.e. both S_ and S1_ would demangle to Hello. This situation should never
+         * arise. A name can always be correctly encoded without repeats. In the above example that
+         * would be _ZN5Hello9compareToEJiPS_.
+         */
         final NativeImageBFDNameProvider nameProvider;
         final StringBuilder sb;
-        final List<String> prefixes;
+
+        // A list of lookup names identifying substituted elements. A prospective name for an
+        // element that is about to be encoded can be looked up in this list. If a match is found
+        // the list index can be used to identify the relevant short symbol. If it is not found
+        // then inserting the name serves to allocate the short name associated with the inserted
+        // element's result index.
+        List<LookupName> bindings;
```

Comment:
As already mentioned. I would prefer a map based approach for recording and lookup of current state of bindings but if @zakkak measurements show that it does not matter in practice I'm fine wit that.

Reply:
I decided to use a map here anyway.